File: lsp_examples/teste.py
import heapq


def dijkstra(graph, start):
    distances = {node: float("inf") for node in graph}
    distances[start] = 0
    priority_queue = [(0, start)]

    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)

        if current_distance > distances[current_node]:
            continue

        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))

    return distances


if __name__ == "__main__":
    graph = {
        "A": {"B": 4, "H": 8},
        "B": {"A": 4, "C": 8, "H": 11},
        "C": {"B": 8, "D": 7, "F": 4},
        "D": {"C": 7, "E": 9, "F": 14},
        "E": {"D": 9, "F": 10},
        "F": {"C": 4, "D": 14, "E": 10, "G": 2},
        "G": {"F": 2, "H": 1, "I": 6},
        "H": {"A": 8, "B": 11, "G": 1, "I": 7},
        "I": {"G": 6, "H": 7},
    }

    start_node = "A"
    shortest_distances = dijkstra(graph, start_node)
    print(f"Shortest distances from node {start_node}: {shortest_distances}")

File: lsp_examples/teste.go
package main

import (
	"container/heap"
	"fmt"
)

type Node struct {
	id   int
	cost int
}

type PriorityQueue []*Node

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].cost < pq[j].cost
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

func (pq *PriorityQueue) Push(x interface{}) {
	item := x.(*Node)
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	item := old[n-1]
	*pq = old[0 : n-1]
	return item
}

func dijkstra(graph [][]int, start int) []int {
	n := len(graph)
	dist := make([]int, n)
	visited := make([]bool, n)

	for i := range dist {
		dist[i] = 1 << 31 // Set distances to maximum integer value
	}

	dist[start] = 0
	pq := make(PriorityQueue, 0)
	heap.Push(&pq, &Node{id: start, cost: 0})

	for pq.Len() > 0 {
		u := heap.Pop(&pq).(*Node)
		uID := u.id
		visited[uID] = true

		for v := 0; v < n; v++ {
			if graph[uID][v] != 0 && !visited[v] {
				newCost := dist[uID] + graph[uID][v]
				if newCost < dist[v] {
					dist[v] = newCost
					heap.Push(&pq, &Node{id: v, cost: newCost})
				}
			}
		}
	}

	return dist
}

func main() {
	graph := [][]int{
		{0, 4, 0, 0, 0, 0, 0, 8, 0},
		{4, 0, 8, 0, 0, 0, 0, 11, 0},
		{0, 8, 0, 7, 0, 4, 0, 0, 2},
		{0, 0, 7, 0, 9, 14, 0, 0, 0},
		{0, 0, 0, 9, 0, 10, 0, 0, 0},
		{0, 0, 4, 14, 10, 0, 2, 0, 0},
		{0, 0, 0, 0, 0, 2, 0, 1, 6},
		{8, 11, 0, 0, 0, 0, 1, 0, 7},
		{0, 0, 2, 0, 0, 0, 6, 7, 0},
	}

	startNode := 0
	shortestDistances := dijkstra(graph, startNode)

	fmt.Printf("Shortest distances from node %d: %v\n", startNode, shortestDistances)
}
